### ISSUE TITLE
Thread safety issue

### ISSUE THREAD

**Issue by arjun27** [pynamodb/PynamoDB#153]
While doing a `put_item` in a table, we're frequently getting an key error on `endpoint_resolver`. The same key error can be seen for `data_loader` and `credential_provider`, but `endpoint_resolver` is far more common.

```
Traceback (most recent call last): 
File "./instrumentation/utils.py", line 69, in save_driver_log 
  driver_log.save() 
File "/app/.heroku/python/lib/python2.7/site-packages/pynamodb/models.py", line 379, in save 
  data = self._get_connection().put_item(*args, **kwargs) 
File "/app/.heroku/python/lib/python2.7/site-packages/pynamodb/connection/table.py", line 83, in put_item 
  return_item_collection_metrics=return_item_collection_metrics) 
File "/app/.heroku/python/lib/python2.7/site-packages/pynamodb/connection/base.py", line 806, in put_item 
  operation_kwargs.update(self.get_identifier_map(table_name, hash_key, range_key, key=ITEM)) 
File "/app/.heroku/python/lib/python2.7/site-packages/pynamodb/connection/base.py", line 648, in get_identifier_map 
  tbl = self.get_meta_table(table_name) 
File "/app/.heroku/python/lib/python2.7/site-packages/pynamodb/connection/base.py", line 399, in get_meta_table 
  data = self.dispatch(DESCRIBE_TABLE, operation_kwargs) 
File "/app/.heroku/python/lib/python2.7/site-packages/pynamodb/connection/base.py", line 235, in dispatch 
  data = self._make_api_call(operation_name, operation_kwargs) 
File "/app/.heroku/python/lib/python2.7/site-packages/pynamodb/connection/base.py", line 250, in _make_api_call 
  operation_model = self.client._service_model.operation_model(operation_name) 
File "/app/.heroku/python/lib/python2.7/site-packages/pynamodb/connection/base.py", line 387, in client 
  self._client = self.session.create_client(SERVICE_NAME, self.region, endpoint_url=self.host) 
File "/app/.heroku/python/lib/python2.7/site-packages/botocore/session.py", line 809, in create_client 
  endpoint_resolver = self.get_component('endpoint_resolver') 
File "/app/.heroku/python/lib/python2.7/site-packages/botocore/session.py", line 690, in get_component 
  return self._components.get_component(name) 
File "/app/.heroku/python/lib/python2.7/site-packages/botocore/session.py", line 875, in get_component 
    del self._deferred[name] 
KeyError: 'endpoint_resolver' 
```

This seems to be related to thread safety in botocore (reference: https://geekpete.com/blog/multithreading-boto3/)

The boto3 documentation recommends creating a new session for each thread (reference: http://boto3.readthedocs.io/en/latest/guide/resources.html#multithreading) which I believe is done with `get_session()` in `base.py#L365`.

What am I missing? Any hints will be very helpful

**Comment by victoreufrasio** [comment]
To reproduce this error, you can:

1. Declare a model

```python
from pynamodb.models import Model
from pynamodb.attributes import UnicodeAttribute

class Test(Model):
	class Meta:
		table_name = 'test'
		region = "us-east-1"

	partition = UnicodeAttribute(hash_key = True)
	sort = UnicodeAttribute(range_key = True)
```

2. Create multiple threads using that model 

```python
from threading import Thread
from pymodels.test import Test

def putItem():
	while True:
		item = Test(
			partition = "test",
			sort = 'test'
		)
		item.save()

for i in range(50):
	t = Thread(target=putItem)
	t.daemon = False
	t.start()
```

**Comment by jpinner-lyft** [comment]
@arjun27 I'm having difficulty reproducing this issue -- what version of botocore are you using?
```
>>> import botocore
>>> botocore.__version__
'1.2.0'
```

**Comment by jpinner-lyft** [comment]
Looks like this change introduced the issue:
https://github.com/boto/botocore/commit/1ed7846b8da056ca8191fc10552d604df618b064

**Comment by farshidz** [comment]
Is this fix going to be released soon? Using 3.2.1 and having the same issue here
